File: apps/evobase_workbench/lib/client/models/docs.dart
import 'package:json_annotation/json_annotation.dart';

import 'databases.dart';

part 'docs.g.dart';

/// Mirror of evobase-protocol ApiDocsDto
@JsonSerializable()
class ApiDocsDto {
  final DatabaseDto database;
  final List<TableDocDto> tables;

  ApiDocsDto({required this.database, required this.tables});

  factory ApiDocsDto.fromJson(Map<String, dynamic> json) =>
      _$ApiDocsDtoFromJson(json);

  Map<String, dynamic> toJson() => _$ApiDocsDtoToJson(this);
}

@JsonSerializable()
class TableDocDto {
  final String name;
  final String schema;
  final String table;
  final String endpoint;
  final List<ColumnDocDto> columns;
  final TableMethodsDto methods;
  final RlsDocDto rls;
  final QueryDocDto query;

  TableDocDto({
    required this.name,
    required this.schema,
    required this.table,
    required this.endpoint,
    required this.columns,
    required this.methods,
    required this.rls,
    required this.query,
  });

  factory TableDocDto.fromJson(Map<String, dynamic> json) =>
      _$TableDocDtoFromJson(json);

  Map<String, dynamic> toJson() => _$TableDocDtoToJson(this);
}

@JsonSerializable()
class ColumnDocDto {
  final String name;

  @JsonKey(name: 'type')
  final String dataType;

  final bool nullable;

  @JsonKey(name: 'has_default')
  final bool hasDefault;

  ColumnDocDto({
    required this.name,
    required this.dataType,
    required this.nullable,
    required this.hasDefault,
  });

  factory ColumnDocDto.fromJson(Map<String, dynamic> json) =>
      _$ColumnDocDtoFromJson(json);

  Map<String, dynamic> toJson() => _$ColumnDocDtoToJson(this);
}

@JsonSerializable()
class RlsDocDto {
  final bool enabled;
  final List<RlsPolicyDocDto> policies;

  RlsDocDto({required this.enabled, required this.policies});

  factory RlsDocDto.fromJson(Map<String, dynamic> json) =>
      _$RlsDocDtoFromJson(json);

  Map<String, dynamic> toJson() => _$RlsDocDtoToJson(this);
}

@JsonSerializable()
class RlsPolicyDocDto {
  final String name;
  final String command;

  @JsonKey(name: 'using')
  final String? using;

  @JsonKey(name: 'with_check')
  final String? withCheck;

  RlsPolicyDocDto({
    required this.name,
    required this.command,
    this.using,
    this.withCheck,
  });

  factory RlsPolicyDocDto.fromJson(Map<String, dynamic> json) =>
      _$RlsPolicyDocDtoFromJson(json);

  Map<String, dynamic> toJson() => _$RlsPolicyDocDtoToJson(this);
}

@JsonSerializable()
class TableMethodsDto {
  @JsonKey(name: 'GET')
  final bool get;

  @JsonKey(name: 'POST')
  final bool post;

  @JsonKey(name: 'PATCH')
  final bool patch;

  @JsonKey(name: 'DELETE')
  final bool delete;

  TableMethodsDto({
    required this.get,
    required this.post,
    required this.patch,
    required this.delete,
  });

  factory TableMethodsDto.fromJson(Map<String, dynamic> json) =>
      _$TableMethodsDtoFromJson(json);

  Map<String, dynamic> toJson() => _$TableMethodsDtoToJson(this);
}

@JsonSerializable()
class QueryDocDto {
  @JsonKey(name: 'selectable_columns')
  final List<String> selectableColumns;

  @JsonKey(name: 'filter_operators')
  final List<String> filterOperators;

  @JsonKey(name: 'order_supported')
  final bool orderSupported;

  @JsonKey(name: 'limit_supported')
  final bool limitSupported;

  @JsonKey(name: 'offset_supported')
  final bool offsetSupported;

  QueryDocDto({
    required this.selectableColumns,
    required this.filterOperators,
    required this.orderSupported,
    required this.limitSupported,
    required this.offsetSupported,
  });

  factory QueryDocDto.fromJson(Map<String, dynamic> json) =>
      _$QueryDocDtoFromJson(json);

  Map<String, dynamic> toJson() => _$QueryDocDtoToJson(this);
}

File: crates/evobase-protocol/src/docs.rs
use serde::Serialize;

use crate::DatabaseDto;

#[derive(Debug, Clone, Serialize)]
pub struct ApiDocsDto {
    pub database: DatabaseDto,
    pub tables: Vec<TableDocDto>,
}

#[derive(Debug, Clone, Serialize)]
pub struct TableDocDto {
    pub name: String,
    pub schema: String,
    pub table: String,
    pub endpoint: String,
    pub columns: Vec<ColumnDocDto>,
    pub methods: TableMethodsDto,
    pub rls: RlsDocDto,
    pub query: QueryDocDto,
}

#[derive(Debug, Clone, Serialize)]
pub struct ColumnDocDto {
    pub name: String,
    #[serde(rename = "type")]
    pub data_type: String,
    pub nullable: bool,
    pub has_default: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct RlsDocDto {
    pub enabled: bool,
    pub policies: Vec<RlsPolicyDocDto>,
}

#[derive(Debug, Clone, Serialize)]
pub struct RlsPolicyDocDto {
    pub name: String,
    pub command: String,
    #[serde(rename = "using", skip_serializing_if = "Option::is_none")]
    pub r#using: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub with_check: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
pub struct TableMethodsDto {
    #[serde(rename = "GET")]
    pub get: bool,
    #[serde(rename = "POST")]
    pub post: bool,
    #[serde(rename = "PATCH")]
    pub patch: bool,
    #[serde(rename = "DELETE")]
    pub delete: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct QueryDocDto {
    pub selectable_columns: Vec<String>,
    pub filter_operators: Vec<&'static str>,
    pub order_supported: bool,
    pub limit_supported: bool,
    pub offset_supported: bool,
}

impl ApiDocsDto {
    pub fn new(database: DatabaseDto, tables: Vec<TableDocDto>) -> Self {
        Self { database, tables }
    }
}

impl From<evobase_core::TableDoc> for TableDocDto {
    fn from(doc: evobase_core::TableDoc) -> Self {
        Self {
            name: doc.name,
            schema: doc.schema,
            table: doc.table,
            endpoint: doc.endpoint,
            columns: doc.columns.into_iter().map(Into::into).collect(),
            methods: doc.methods.into(),
            rls: doc.rls.into(),
            query: doc.query.into(),
        }
    }
}

impl From<evobase_core::ColumnDoc> for ColumnDocDto {
    fn from(col: evobase_core::ColumnDoc) -> Self {
        Self {
            name: col.name,
            data_type: col.data_type,
            nullable: col.nullable,
            has_default: col.has_default,
        }
    }
}

impl From<evobase_core::RlsDoc> for RlsDocDto {
    fn from(rls: evobase_core::RlsDoc) -> Self {
        Self {
            enabled: rls.enabled,
            policies: rls.policies.into_iter().map(Into::into).collect(),
        }
    }
}

impl From<evobase_core::RlsPolicyDoc> for RlsPolicyDocDto {
    fn from(policy: evobase_core::RlsPolicyDoc) -> Self {
        Self {
            name: policy.name,
            command: policy.command,
            r#using: policy.r#using,
            with_check: policy.with_check,
        }
    }
}

impl From<evobase_core::TableMethods> for TableMethodsDto {
    fn from(methods: evobase_core::TableMethods) -> Self {
        Self {
            get: methods.get,
            post: methods.post,
            patch: methods.patch,
            delete: methods.delete,
        }
    }
}

impl From<evobase_core::QueryDoc> for QueryDocDto {
    fn from(query: evobase_core::QueryDoc) -> Self {
        Self {
            selectable_columns: query.selectable_columns,
            filter_operators: query.filter_operators,
            order_supported: query.order_supported,
            limit_supported: query.limit_supported,
            offset_supported: query.offset_supported,
        }
    }
}
